Optimizer: improve the load-copy-to-borrow optimization and implement it in swift

eeckstein opened the pull request:
The optimization replaces a `load [copy]` with a `load_borrow` if possible.

```
  %1 = load [copy] %0
  // no writes to %0
  destroy_value %1
```
->
```
  %1 = load_borrow %0
  // no writes to %0
  end_borrow %1
```

The new implementation uses alias-analysis (instead of a simple def-use walk), which is much more powerful.
Also, implement load-borrow-immutability checkin in the swift verifier.

rdar://115315849

eeckstein committed:
Swift SIL: add `Operand.changeOwnership`

eeckstein committed:
Swift SIL: add flag parameters to `Builder.createBeginBorrow`

eeckstein committed:
Swift Optimizer: add `SpecificInstructionSet` and `SpecificInstructionWorklist`

To be used for specific instruction types

eeckstein committed:
SwiftCompilerSources: add the `LoadBorrowInst.isUnchecked` API

eeckstein committed:
SIL Verifier: implement load-borrow-immutability checkin in the swift verifier

eeckstein committed:
Optimizer: add the `FunctionPassContext.completeLifetime(of: Value)` utility

Implemented by bridging the OSSALifetimeCompletion utility

eeckstein committed:
Optimizer: improve the load-copy-to-borrow optimization and implement it in swift

The optimization replaces a `load [copy]` with a `load_borrow` if possible.

```
  %1 = load [copy] %0
  // no writes to %0
  destroy_value %1
```
->
```
  %1 = load_borrow %0
  // no writes to %0
  end_borrow %1
```

The new implementation uses alias-analysis (instead of a simple def-use walk), which is much more powerful.

rdar://115315849

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci benchmark

eeckstein commented:
@swift-ci test

nate-chandler commented on SwiftCompilerSources/Sources/Optimizer/PassManager/Context.swift:
```diff
@@ -422,6 +422,14 @@ struct FunctionPassContext : MutatingContext {
+
+  /// Makes sure that the lifetime of `value` ends at all control flow paths, even in dead-end blocks.
+  /// Inserts destroys in dead-end blocks if those are missing.
+  func completeLifetime(of value: Value) {
```
Is now a good time to add a parameter to control whether lifetimes end at the availability or lifetime boundary?

nate-chandler commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift:
```diff
@@ -0,0 +1,264 @@
+    return
+  }
+
+  var liferange = InstructionWorklist(context)
```
Nit: `liferange` -> `liverange` (or `lifetime`) throughout.

nate-chandler commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift:
```diff
@@ -0,0 +1,264 @@
+  //     ...          // no destroy of %1!
+  //   unreachable    // the lifetime of %1 ends here
+  //
+  context.completeLifetime(of: loadedValue)
```
Generally, lifetimes need to be completed inside out/bottom up (as in `SILGenCleanup::completeOSSALifetimes`).  I haven't come up with an example where that would be a problem in this usage, but I'm not confident it's not.

nate-chandler commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift:
```diff
@@ -0,0 +1,264 @@
+    return true
+  }
+
+  private mutating func findNonDestroyingLiferangeExists(of forwardingInst: ForwardingInstruction) {
```
`findNonDestroyingLiferangeExists` -> `findNonDestroyingLiferangeExits`

nate-chandler commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift:
```diff
@@ -0,0 +1,264 @@
+
+  // Handle the special case if the `load` is immediately followed by a single `move_value`.
+  // Note that `move_value` is _not_ a forwarding instruction.
+  // In this case we have to preserve the move's flags by inserting a `begin_borrow` with the same flags.
```
Would this portion of the comment be most useful in `replaceWithLoadBorrow(liferange:ownershipUses:)` near
```
     if let moveInst = singleMoveValueUser {
```
?

eeckstein commented:
@swift-ci test macos

nate-chandler reviewed (approved):
LGTM, modulo the question of completing lifetimes inside out.

eeckstein committed:
OwnershipLiveness: add a flag `visitInnerUses` to enable constructing incomplete liveranges

If `visitInnerUses ` is true, it's not assumed that inner scopes are linear.
It forces to visit all interior uses if inner scopes.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

atrick reviewed (commented):
Observations and future suggestions...

--- Liverange::collectUses ---

`InteriorUseWalker::init` does not actually need `visitInnerUses: Bool` because it could just have a `visitInnerUses` entry point that does:
```
innerScopeHandler = { return visitUsesOfOuter(value: $0) }
```

`Liverange::collectUses` is just doing what I would call "forwarding-liveness". We could standardize that: basically a combination of `computeLinearLiveness` and `ForwardingDefUseWalker`.

`findAdditionalUsesInDeadEndBlocks` can be eliminated with complete OSSA... ok, I just noticed a TODO says the same thing. But until then, it seems like it might be as efficient to call `completeLife(of:)` first for each owned values, then simply look for `endsLifetime` uses. Note that you do need to handle `ExtendLifetimeInst`, just like `computeLinearLiveness` does.

--- LoadBorrowInst::verify ---

Small note on load-borrow-immutability verification:

At some point, I would like to centralize the definition of legal address uses in one place. So `isMutatedAddress` would be alongside `AddressUseVisitor::classifyAddress`. Also note that, unlike the incorrectly named `AddressDefUseWalker::leafUse`, `AddressUseVisitor::leafAddressUse` is actually a leaf use, so you could probably just check `mayWrite` instead of switching over opcodes.

Then verification should be improved so it does not ignore "unknown" uses (although in this particular verification, we always need to ignore *known escaping* uses).

eeckstein commented:
> InteriorUseWalker::init does not actually need visitInnerUses: Bool because it could just have a visitInnerUses entry point that does:

👍. Though the drawback is that this is an escaping closure and needs memory allocation for its context. It's potentially called many times (for each load). So this might affect compile time.

> so you could probably just check mayWrite instead of switching over opcodes.

That doesn't work because `mayWrite` is less precise than what the load-copy-to-borrow optimization does using alias analysis. It would result in false alarms.
